java/query/builder/Sortable.java

```diff
@@ -67,43 +73,49 @@ default S sort(List<UnboundVariable> vars, TypeQLArg.Order order) {
 
     class Sorting {
 
-        private final List<UnboundVariable> vars;
-        private final TypeQLArg.Order order;
         private final int hash;
+        private final List<UnboundVariable> variables;
+        private final Map<UnboundVariable, TypeQLArg.Order> orders;
 
-        public Sorting(List<UnboundVariable> vars) {
-            this(vars, null);
+        private Sorting(List<UnboundVariable> variables, Map<UnboundVariable, TypeQLArg.Order> orders) {
+            this.variables = variables;
+            this.orders = orders;
+            this.hash = Objects.hash(variables, orders);
         }
 
-        public Sorting(List<UnboundVariable> vars, TypeQLArg.Order order) {
-            this.vars = vars;
-            this.order = order;
-            this.hash = Objects.hash(vars(), order());
+        public static Sorting create(List<Pair<UnboundVariable, TypeQLArg.Order>> sorting) {
+            List<UnboundVariable> vars = new ArrayList<>();
+            Map<UnboundVariable, TypeQLArg.Order> orders = new HashMap<>();
+            sorting.forEach(pair -> {
+                vars.add(pair.first());
+                orders.put(pair.first(), pair.second());
+            });
+            return new Sorting(vars, orders);
         }
 
-        public List<UnboundVariable> vars() {
-            return vars;
+        public List<UnboundVariable> variables() {
+            return variables;
         }
 
-        public TypeQLArg.Order order() {
-            return order == null ? TypeQLArg.Order.ASC : order;
+        public TypeQLArg.Order getOrder(UnboundVariable var) {
+            if (variables.contains(var)) throw TypeQLException.of(VARIABLE_NOT_SORTED.message(var));
+            return orders.getOrDefault(var, TypeQLArg.Order.ASC);
```

Comment:
we retain `null` in the order map to be able to remember which variables used the default order. Only when retrieving the order do we default it to `ASC`,  but we don't store `ASC` in the map as it would distort the `toString()`